File: cdm/cpp/properties/SEScalarArea.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarArea.h"

const AreaUnit AreaUnit::cm2("cm^2");
const AreaUnit AreaUnit::m2("m^2");

bool AreaUnit::IsValidUnit(const std::string& unit)
{
  if (cm2.GetString().compare(unit) == 0)
    return true;
  if (m2.GetString().compare(unit) == 0)
    return true;
  return false;
}

const AreaUnit& AreaUnit::GetCompoundUnit(const std::string& unit)
{
  if (cm2.GetString().compare(unit) == 0)
    return cm2;
  if (m2.GetString().compare(unit) == 0)
    return m2;
  std::stringstream err;
  err << unit << " is not a valid Area unit";
  throw CommonDataModelException(err.str());
}

void SEScalarArea::Load(const cdm::ScalarAreaData& src, SEScalarArea& dst)
{
  SEScalarArea::Serialize(src, dst);
}
void SEScalarArea::Serialize(const cdm::ScalarAreaData& src, SEScalarArea& dst)
{
  SEScalarQuantity<AreaUnit>::Serialize(src.scalararea(), dst);
}

cdm::ScalarAreaData* SEScalarArea::Unload(const SEScalarArea& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarAreaData* dst = new cdm::ScalarAreaData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarArea::Serialize(const SEScalarArea& src, cdm::ScalarAreaData& dst)
{
  SEScalarQuantity<AreaUnit>::Serialize(src, *dst.mutable_scalararea());
}

File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarArea.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarAreaData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.AreaUnit;

/**
 * @author abray
 * Area Scalar Class, will enforce that units are proper to Areas
 */

public class SEScalarArea extends SEScalar
{
  public SEScalarArea()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarArea(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarArea(double value, AreaUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarAreaData src, SEScalarArea dst)
  {
    SEScalar.load(src.getScalarArea(),dst);
  }
  public static ScalarAreaData unload(SEScalarArea src)
  {
    if(!src.isValid())
      return null;
    ScalarAreaData.Builder dst = ScalarAreaData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarArea src, ScalarAreaData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarAreaBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, AreaUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(AreaUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(AreaUnit.validUnit(unit))
      return true;
    return false;
  }
}
